rotkehlchen/externalapis/alchemy.py

```diff
@@ -0,0 +1,400 @@
+        """Deserializes historical price data from Alchemy API response.
+
+        If price data is missing, invalid, or there's an error, ZERO_PRICE is returned.
+        Otherwise, the deserialized price is returned.
+        """
+        try:
+            if self._is_identifier_match(result, asset_identifier) is False:
+                log.warning(
+                    f'Queried Alchemy API from {from_asset.identifier} to {to_asset.identifier}, '
+                    f'but received mismatched symbol in response',
+                )
+                return ZERO_PRICE
+
+            if result['currency'] != 'usd':
+                log.warning(
+                    f'Queried Alchemy API from {from_asset.identifier} to {to_asset.identifier}, '
+                    f'but received non-USD currency in response',
+                )
+                return ZERO_PRICE
+
+            if not (data := result.get('data', [])):
+                log.warning(
+                    f'Queried Alchemy API from {from_asset.identifier} to {to_asset.identifier}, '
+                    f'but received no historical price data in response',
+                )
+                return ZERO_PRICE
+
+            closest_price = min(
+                data,
+                key=lambda x: abs(timestamp - int(datetime.fromisoformat(x['timestamp']).timestamp())),  # noqa: E501
+            )
+            return deserialize_price(closest_price['value'])
+        except (DeserializationError, KeyError, ValueError) as e:
+            error_msg = f'missing key {e!s}' if isinstance(e, KeyError) else str(e)
+            log.warning(
+                f'Queried Alchemy Price API from {from_asset.identifier} to {to_asset.identifier}. '  # noqa: E501
+                f'Failed to deserialize historical price data due to: {error_msg}',
+            )
+            return ZERO_PRICE
+
+    @staticmethod
+    def _is_identifier_match(
+            entry: dict[str, Any],
+            asset_identifier: dict[str, str],
+    ) -> bool:
+        """Check if the API response entry matches the asset identifier."""
+        return (
+            ('symbol' in asset_identifier and entry['symbol'] == asset_identifier['symbol']) or
+            (
+                'address' in asset_identifier and
+                entry['network'] == asset_identifier['network'] and
+                entry['address'] == asset_identifier['address']
+            )
+        )
+
+    def _query(
+            self,
+            payload: dict[str, Any],
+            endpoint: Literal['by-symbol', 'by-address', 'historical'],
+    ) -> dict[str, Any]:
+        """
+        May raise:
+        - RemoteError
+        """
+        if (api_key := self._get_api_key()) is None:
+            api_key = self.default_api_key
+
+        url = f'https://api.g.alchemy.com/prices/v1/{api_key}/tokens/{endpoint}'
+        log.debug(f'Querying alchemy api: {url=} with {payload=}')
+        try:
+            if endpoint == 'by-symbol':
```

Comment:
no need to have this if check in every query. Just put the verb and the params as part of the `_query()` function. It's not used in that many places anyway and will save you an if check in every single API call